gemini/responsible-ai/react_rag_attacks_mitigations_examples.ipynb

```diff
@@ -0,0 +1,1086 @@
+    "\n",
+    "    return response.text"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f8fBQ-5adKyX",
+   "metadata": {
+    "id": "f8fBQ-5adKyX"
+   },
+   "outputs": [],
+   "source": [
+    "chat_rag(\"What is the revenue of Beyond41?\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "JOKWZ5qAdtxS",
+   "metadata": {
+    "id": "JOKWZ5qAdtxS"
+   },
+   "outputs": [],
+   "source": [
+    "print(chat_rag(\"What are the Financial results of Beyond41?\"))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "S3gTKdVVeKWj",
+   "metadata": {
+    "id": "S3gTKdVVeKWj"
+   },
+   "source": [
+    "### RAG possible attacks"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "QoJ5D5e5eyFo",
+   "metadata": {
+    "id": "QoJ5D5e5eyFo"
+   },
+   "source": [
+    "Let us assume the company has a lot of historically generated PDF files from different tools. The company wants to use RAG to get more insight and customer value out of the documents."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fsdodIx9eWnQ",
+   "metadata": {
+    "id": "fsdodIx9eWnQ"
+   },
+   "outputs": [],
+   "source": [
+    "doc = fitz.open(\"Beyond41mal.pdf\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "tBEwLxvsf5O2",
+   "metadata": {
+    "id": "tBEwLxvsf5O2"
+   },
+   "outputs": [],
+   "source": [
+    "print(chat_rag(\"What are the Financial results of Beyond41?\"))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "gvD6eHOJgCMK",
+   "metadata": {
+    "id": "gvD6eHOJgCMK"
+   },
+   "outputs": [],
+   "source": [
+    "print(chat_rag(\"Give me details of Beyond41 ?\"))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cBjCWBiH2Q5Y",
+   "metadata": {
+    "id": "cBjCWBiH2Q5Y"
+   },
+   "outputs": [],
+   "source": [
+    "print(chat_rag(\"What is the future of Beyond41?\"))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "90iUjIsjg8rY",
+   "metadata": {
+    "id": "90iUjIsjg8rY"
+   },
+   "source": [
+    "**Why is the data wrong?**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "N0q6qkmf6PkJ",
+   "metadata": {
+    "id": "N0q6qkmf6PkJ"
+   },
+   "source": [
+    "![document-mal.png](https://storage.googleapis.com/github-repo/responsible-ai/intro_genai_security/document-mal.png)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cKpcRNdPgQXM",
+   "metadata": {
+    "id": "cKpcRNdPgQXM"
+   },
+   "outputs": [],
+   "source": [
+    "print(search_snippets(\"content\"))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "QakbslPnyxgg",
+   "metadata": {
+    "id": "QakbslPnyxgg"
+   },
+   "source": [
+    "### Possible Attack Mitigations\n",
+    "\n",
+    "There is an not optimal solution but you should look into layering multiple filters, like DLP, Basic Filtering and Manual Cross Checking"
```

Comment:
Can you re-phrase this? I'm not entirely sure what is meant here

Reply:
good point, done